src/server.rs

```diff
@@ -0,0 +1,475 @@
+use std::collections::hash_map::Entry;
+use std::collections::VecDeque;
+use std::fmt;
+use std::mem::take;
+use std::sync::Arc;
+use std::task::{ready, Context, Poll};
+
+use asynchronous_codec::FramedWrite;
+use blockstore::{Blockstore, BlockstoreError};
+use cid::CidGeneric;
+use fnv::{FnvHashMap, FnvHashSet};
+use futures::future::{AbortHandle, Abortable, BoxFuture};
+use futures::stream::FuturesUnordered;
+use futures::{FutureExt, SinkExt, StreamExt};
+use libp2p_core::upgrade::ReadyUpgrade;
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    ConnectionHandlerEvent, NotifyHandler, StreamProtocol, SubstreamProtocol, ToSwarm,
+};
+use smallvec::SmallVec;
+use tracing::{debug, info, trace, warn};
+
+use crate::incoming_stream::ServerMessage;
+use crate::message::Codec;
+use crate::proto::message::{
+    mod_Message::Block as ProtoBlock, mod_Message::Wantlist as ProtoWantlist, Message,
+};
+use crate::utils::stream_protocol;
+use crate::{Event, Result, StreamRequester, ToBehaviourEvent, ToHandlerEvent};
+
+type Sink = FramedWrite<libp2p_swarm::Stream, Codec>;
+type BlockWithCid<const S: usize> = (CidGeneric<S>, Vec<u8>);
+
+#[derive(Debug)]
+pub(crate) struct ServerBehaviour<const S: usize, B>
+where
+    B: Blockstore + Send + Send,
+{
+    protocol: StreamProtocol,
+    store: Arc<B>,
+
+    peers_waitlists: FnvHashMap<PeerId, PeerWantlist<S>>,
```

Comment:
I'm wondering if wrapping PeerId in Arc would be beneficial as we keep a few for each peer here and there

Reply:
I thought it would cloud the code more, but it doesn't look that bad - 798d799b9b38150ddf01ea32556164c3274c4bc5

Reply:
indeed, so wdyt? @oblique? is it worth it

Reply:
So.. `PeerId` is 80 bytes does not allocate in heap ([1](https://github.com/libp2p/rust-libp2p/blob/7a3101d83f5a75da03fe2cf580862a9c1fc910c1/identity/src/peer_id.rs#L50), [2](https://github.com/multiformats/rust-multihash/blob/ef4e92a418005d5c27ae417477606bf705eb8556/src/multihash.rs#L39-L44)) and `HashMap` stores the key and value in memory in [as tuple](https://github.com/rust-lang/hashbrown/blob/274c7bbd79398881e0225c0133e423ce60d7a8f1/src/map.rs#L192). In case of `Arc` we copy less bytes but we fragment the memory because of heap. I would choose non-Arc in our case.